File: Android/kwebtonative/src/main/java/com/grootan/kwebtonative/KWebToNative.java
package com.grootan.kwebtonative;

import android.webkit.WebView;
import android.webkit.WebViewClient;
import com.google.gson.Gson;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

/**
 * Main Singleton Class to handle webview wrapper
 */
public class KWebToNative {

    public static KWebToNative _shared;
    private HashMap<String, List<KWebToNativeListenerDelegate>> _listeners;
    private WebView _webview;
    private Gson _gson;
    private KWebToNativeWebViewClient _client;

    /**
     * Private constructor
     */
    private KWebToNative() {
        _listeners = new HashMap<>();
        _gson = new Gson();
    }

    /**
     * Method to get instance of KWebToNative
     * @return
     */
    public static KWebToNative shared() {
        if (_shared == null) {
            _shared = new KWebToNative();
        }
        return _shared;
    }

    /**
     * Method to initialize webview and webviewclient
     * @param webView
     * @param webViewClient
     */
    public void initialize(WebView webView, WebViewClient webViewClient)
    {
        if(_client == null)
        {
            _client = new KWebToNativeWebViewClient(_shared);
        }
        this._client.setDelegate(webViewClient);
        this._webview = webView;
        webView.getSettings().setJavaScriptEnabled(true);
        webView.setWebViewClient(_client);
    }

    /**
     * Method to listen on events
     * @param eventName
     * @param delegate
     */
    public void on(String eventName,KWebToNativeListenerDelegate delegate )
    {
        List<KWebToNativeListenerDelegate> delegates = _listeners.get(eventName);
        if(delegates == null)
        {
            delegates = new ArrayList<>();
        }
        delegates.add(delegate);
        _listeners.put(eventName,delegates);
    }

    /**
     * Method not to listen for events, it will remove the listener which is called with on
     * @param eventName
     */
    public void off(String eventName)
    {
        _listeners.remove(eventName);
    }

    /**
     * Method to send payload to webview
     * @param eventName
     * @param payload
     */
    public void send(String eventName, Object payload)
    {
        send(eventName, payload);
    }

    /**
     * Method to send payload to webview with callback
     * @param eventName
     * @param payload
     * @param delegate
     */
    public void send(String eventName, Object payload, KWebToNativeCallbackDelegate delegate)
    {
        if(_webview == null)
        {
            throw new NullPointerException("Webview cannot be null, you must call initialize before calling send");
        }

        String jsonData = _gson.toJson(payload);
        _webview.loadUrl(String.format("javascript:KWebToNative.trigger(\"%s\",%s);",eventName,jsonData));
        delegate.onComplete();
    }

    /**
     * Method to handled payload from webview
     * @param webView
     * @param envelope
     */
    protected void triggerEventFromWebView(final WebView webView,
                                           WebViewPayload envelope) {
        String type = envelope.type;

        if (_listeners.containsKey(type)) {
            List<KWebToNativeListenerDelegate> handlers = _listeners.get(type);

            for (KWebToNativeListenerDelegate handler:
                handlers ) {
                handler.onMessage(envelope.payload);
            }
        }
    }


}

File: iOS/KWebToNative/KWebToNative.swift


import Foundation
import WebKit

/**
 Delegate for listeners
 */
public protocol KWebToNativeCallBackDelegate: class {
    func onComplete(_ message:Any, _ error: Error)
}

/**
 Delegate for callback
 */
public protocol KWebToNativeListenerDelegate: class {
    func onMessage(_ payload: NSDictionary)
}

/**
 Error enum
 */
enum KWebToNativeError : Error{
    case invalid(String)
}

public class KWebToNative {
    
    public static var shared = KWebToNative()
    var _wkWebView: WKWebView?
    var _listenerDictionary : NSMutableDictionary
    
    /**
     Init to initialize properties of KWebToNative
     */
    private init(){
        _listenerDictionary = NSMutableDictionary()
    }
    
    /**
     Method to set the WKWebview
     */
    public func initialize(_ wkWebView: WKWebView)
    {
        self._wkWebView = wkWebView
    }
    
    /**
     Method to listen for events from WKWebview
     */
    public func on(_ eventName: String,_ listenerDelegate: KWebToNativeListenerDelegate)
    {
        let callBackArray: NSMutableArray = _listenerDictionary.object(forKey: eventName) as? NSMutableArray ?? []
        if(callBackArray == [])
        {
            _listenerDictionary.setValue(callBackArray, forKey: eventName)
        }

        callBackArray.add(listenerDelegate)
    }
    
    /**
     Method to remove events from listening, its always important to remove the events from listening
     */
    public func off(_ eventName: String)
    {
        _listenerDictionary.removeObject(forKey: eventName)
    }
    
    /**
     Method to send message to WkWebview without callback
     */
    public func send(_ eventName: String, _ payload: NSDictionary) throws
    {
        try self.send(eventName, payload, nil)
    }
    
    /**
     Method to send message to WkWebview with callback
     */
    public func send(_ eventName: String, _ payload: NSDictionary,_ callback: KWebToNativeCallBackDelegate?) throws
    {
        guard let webview = _wkWebView else {
            throw KWebToNativeError.invalid("Webview cannot be nil, you must call initialize before calling send")
        }
        
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: payload, options: .prettyPrinted)
            guard let jsonString = String(data: jsonData, encoding:  .utf8) else{
                return
            }
            log("send",String(format: "KWebToNative.trigger(\"%@\",%@);", eventName,jsonString))
            webview.evaluateJavaScript(String(format: "KWebToNative.trigger(\"%@\",%@);", eventName,jsonString)) { (message, error) in
                               
                               if(callback != nil)
                               {
                                   callback!.onComplete(message as Any, error!)
                               }
                           }
        } catch let error as NSError {
            throw error
        }
    }
    
    /**
     Method to process result from webview
     */
    public func process(_ url: URL?)
    {
        if(url == nil)
        {
            return
        }
        
        log("process","Received URL \(String(describing: url))")
        log("process","Scheme \(url?.scheme ?? "No Scheme")")
        if( url?.scheme == "kwebtonative")
        {
            let eventType = url?.host
            let query = url?.query
            let jsonString = query?.removingPercentEncoding
            log("process","Received JSON \(String(describing: jsonString))")
            let jsonData = convertToDictionary(text: jsonString!)
            
            if(eventType == "event")
            {
                triggerEventsFromWeb(jsonData! as NSDictionary)
            }
        
        }
    }
    
    /**
    Method to trigger subscribed listeners
    */
    private func triggerEventsFromWeb(_ jsonData: NSDictionary)
    {
        let eventName = jsonData.object(forKey: "type")
        log("triggerEventsFromWeb","Event Name \(String(describing: eventName))")
        let payload = jsonData.object(forKey: "payload")
        log("triggerEventsFromWeb","Payload \(String(describing: payload))")
        let listeners = _listenerDictionary.object(forKey: eventName as Any) as? NSArray
        for listener in listeners! {
            (listener as! KWebToNativeListenerDelegate).onMessage(payload as! NSDictionary)
        }
    }
    
    /**
    Method to convert jsont to nsdictionary
    */
    private func convertToDictionary(text: String) -> [String: Any]? {
        if let data = text.data(using: .utf8) {
            do {
                return try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any]
            } catch {
                print(error.localizedDescription)
            }
        }
        return nil
    }
    
    /**
     Method to log
     */
    private func log(_ methodName: String, _ message: String)
    {
        print("KWEBTONATIVE:\(methodName):\(message)")
    }
    
}
